### PR TITLE
Export ONNX models with readable input/output names

### PR DESCRIPTION
**Pull request by emcastillo**
For some ONNX exported models, the inputs/outputs names have sometimes a numeric value and this makes pretty hard to inspect the generated graphs in the case of large models.

The solution in this PR was initially submitted to our internal utilities library by @take-cheeze https://github.com/pfnet/pytorch-pfn-extras/pull/102

Now we would like to upstream this change by adding an extra kwarg when exporting the model to allow replacing these numeric names with actual debuggable ones.

As an example, the following code shows that the module output is `3`

```python
g, p, o = _model_to_graph(module, torch.ones(1, 10))
for n in g.nodes():
    for v in n.outputs():
        print(v.debugName())
```
output
```
3
```

With this PR

```
v3_Gemm
```

This allows identifying this out as a value from the associated Gemm layer.

### CODE REVIEW

**Comment by samdow** [comment]
@malfet Do you know who is the best to review/import from the FB side? I see that you've reviewed a couple of ONNX PRs recently so was wondering if it was you

**Comment by malfet** [comment]
@BowenBao or @garymm can you please have a look?

**Comment by garymm** [comment]
@jiafatom is on the torch-ONNX team and can review this. Thanks.

**Comment by emcastillo** [comment]
@jiafatom Thanks a lot for the reply!!!

> https://github.com/pytorch/pytorch/blob/master/torch/onnx/utils.py#L108
That seems more flexible. The way this PR does looks a little ad-hoc.

Yeah, we were looking into using `output_names` but if we understood correctly, `output_names` will rename the graph's final outputs, while our approach renames all the intermediate node outputs inside the graph. We are using this code internally and it really helps us in the debugging process for large graphs.

Nevertheless, I agree that the proposed solution feels like a hack so I am willing to implement it in a cleaner way if you guys can give me some hints :) since I am pretty new to the ONNX codebase.

**Review by garymm — CHANGES_REQUESTED** [review]
I like the idea of this change, but I don't think we should add an option to an API. This seems like it should always be safe to do, so let's just have it on always.

File: torch/onnx/utils.py
Comment (take-cheeze) [reply torch/onnx/utils.py:815]: Some inputs maybe constant propagated initializers which may only have unique name.
So not all inputs are outputs of some other nodes.

File: torch/onnx/utils.py
Comment (take-cheeze) [reply torch/onnx/utils.py:816]: Python is not PHP so comparison of string and integer is always false.
Though I feel `if v.debugName() != str(v.unique())` is more friendly.

File: torch/onnx/utils.py
Comment (garymm) [reply torch/onnx/utils.py:815]: OK, then can we just go over all the inputs and name those after the node that consumes them?
If a node is not an input to anything and not a model output, then I guess it's unused and it doesn't need an interesting name.

**Review by garymm — CHANGES_REQUESTED** [review]

**Comment by emcastillo** [comment]
@garymm thanks for all the comments! I updated the PR accordingly and added a small test.
PTAL!

File: test/onnx/test_utility_funs.py
Comment (garymm) [inline test/onnx/test_utility_funs.py:1152]: Let's do `v.debugName().startswith("onnx::Mul_")`

File: torch/onnx/utils.py
Comment (garymm) [inline torch/onnx/utils.py:808]: Please name this `_apply_friendly_debug_names` or something like that.
(The current name is bad because this doesn't actually do anything to output IDs)

File: torch/onnx/utils.py
Comment (garymm) [inline torch/onnx/utils.py:813]: nit: if you move this above the if, you'll save a call to `v.debugName()`

**Review by garymm — APPROVED** [review]
Thanks, only minor changes requested. I don't work at Meta so I can't merge this but once you've made the requested changes I can nag a Meta employee for you.

**Comment by emcastillo** [comment]
@garymm all suggestions applied! thanks a lot :D

**Comment by garymm** [comment]
@msaroufim could you please import this?

**Comment by emcastillo** [comment]
@garymm @msaroufim I updated the `expect` tests outs to reflect the changes, now everything should be ok :).
Thanks!

**Comment by emcastillo** [comment]
mm the offending tests are not breaking in my local setup :(,
Any clue why?

**Comment by emcastillo** [comment]
@msaroufim all tests passed :)

**Comment by emcastillo** [comment]
@jiafatom friendly ping :)

**Comment by msaroufim** [comment]
Hi! This is kind of a big PR. Can someone briefly explain to me what's going on so I can merge? most of the changes seem to be `.expect` files?

**Comment by garymm** [comment]
Hi @msaroufim. Happy new year!

The `.expect` files contain models in the ONNX format. They're used by tests that run `torch.onnx.export` and compare the output to the `.expect` file. If the output differs, the test fails. Thus someone who changes the output on purpose has to update the files to get the tests to pass.

(I hate this pattern by the way).

This PR changes the output of `torch.onnx.export` in ~all cases, in a way that's not semantically important but makes the models more readable by humans.

Does that help?

**Comment by emcastillo** [comment]
Thank you, everyone!

**Pull request closed without merging**
